### PR TITLE
HDDS-15275. DiskBalancer getIdealUsage handles zero capacity incorrectly.

### PR DESCRIPTION
**Pull request by slfan1989**
## What changes were proposed in this pull request?

This PR makes DiskBalancerVolumeCalculation#getIdealUsage handle empty volume lists and zero total capacity explicitly.

Previously, getIdealUsage directly divided totalEffectiveUsed by totalCapacity. When the volume list was empty, totalCapacity was 0, which caused the method to return NaN. When all volumes had zero capacity, the method could also produce invalid utilization values. These invalid values could propagate to DiskBalancer reporting and balancing calculations.

## What is the link to the Apache JIRA

JIRA: HDDS-15275. DiskBalancer getIdealUsage handles zero capacity incorrectly.

## How was this patch tested?

Add Junit Test.

### CODE REVIEW

**Review by Gargi-jais11 — APPROVED** [review]
Thanks @slfan1989 for the patch. Nice catch! LGTM!

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java [hunk 1]
```diff
@@ -70,13 +70,22 @@ public static List<VolumeFixedUsage> getVolumeUsages(MutableVolumeSet volumeSet,
...
+    if (volumes == null || volumes.isEmpty()) {
+      return 0.0;
+    }
+
     long totalCapacity = 0L, totalEffectiveUsed = 0L;
-    
+
     for (VolumeFixedUsage volumeUsage : volumes) {
       totalCapacity += volumeUsage.getUsage().getCapacity();
       totalEffectiveUsed += volumeUsage.getEffectiveUsed();
```
Comment (szetszwo) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java]: Let's check each volume and check more cases:
```java
    for (VolumeFixedUsage volumeUsage : volumes) {
      final long capacity = volumeUsage.getUsage().getCapacity();
      if (capacity < 0) {
        throw new IllegalArgumentException("Negative capacity = " + capacity
            + ": " + volumeUsage.getVolume());
      }
      final long effectiveUsed = volumeUsage.getEffectiveUsed();
      if (effectiveUsed < 0) {
        throw new IllegalArgumentException("Negative effective used = " + effectiveUsed
            + ": " + volumeUsage.getVolume());
      }
      if (effectiveUsed > capacity) {
        throw new IllegalArgumentException("Effective used = " + effectiveUsed + " > capacity = " + capacity
            + ": " + volumeUsage.getVolume());
      }
      totalCapacity += capacity;
      totalEffectiveUsed += effectiveUsed;
    }
```

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java [hunk 2]
```diff
@@ -70,13 +70,22 @@ public static List<VolumeFixedUsage> getVolumeUsages(MutableVolumeSet volumeSet,
...
     }
-    
+
+    if (totalCapacity <= 0) {
```
Comment (szetszwo) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java]: To be consistent, if (totalCapacity == 0), let's return 0?

**Review by szetszwo — COMMENTED** [review]
@slfan1989 , thanks for working on this!  Please see the comments inlined.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java (hunk 1 above)
Comment (slfan1989) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java]: @szetszwo Thanks for the suggestion! Addressed by validating each volume’s capacity and effective used before calculating ideal usage, and added tests for the new edge cases.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java (hunk 2 above)
Comment (slfan1989) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerVolumeCalculation.java]: Updated. `totalCapacity == 0` now returns `0.0` for consistency with the empty volume case.

**Comment by Gargi-jais11** [comment]
Last patch LGTM, +1.

**Review by szetszwo — APPROVED** [review]
+1  the change looks good.

**Comment by slfan1989** [comment]
@szetszwo @Gargi-jais11 Thanks for the review and for merging the patch!
